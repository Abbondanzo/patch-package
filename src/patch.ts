import * as fs from "fs-extra"

type HunkHeaderLineParseResult =
  | {
      error: true
      message: string
    }
  | {
      error: false
      original: {
        start: number
        length: number
      }
      patched: {
        start: number
        length: number
      }
    }

export function parseHunkHeaderLine(
  headerLine: string,
): HunkHeaderLineParseResult {
  const match = headerLine
    .trim()
    .match(/^@@ -(\d+)(,(\d+))? \+(\d+)(,(\d+))? @@.*/)
  if (!match) {
    return { error: true, message: `Bad header line: '${headerLine}'` }
  }

  return {
    error: false,
    original: {
      start: Number(match[1]),
      length: Number(match[3] || 0),
    },
    patched: {
      start: Number(match[4]),
      length: Number(match[6] || 0),
    },
  }
}

const trimRight = (s: string) => s.replace(/\s+$/, "")

function invertPatchFileLines(lines: string[]): boolean {
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    if (line.startsWith("--- ")) {
      // switch over startPath and endPath
      const other = lines[i + 1]
      if (line.startsWith("--- /dev/null")) {
        lines[i + 1] = "+++ /dev/null"
        lines[i] = "--- a/" + other.slice("+++ b/".length)
      } else {
        lines[i + 1] = "+++ b/" + line.slice("--- a/".length)
        if (other.startsWith("+++ /dev/null")) {
          lines[i] = "--- /dev/null"
        } else {
          lines[i] = "--- a/" + other.slice("+++ b/".length)
        }
      }
      i++
    } else if (line.startsWith("rename from ")) {
      const fromName = line.slice("rename from ".length)
      const toName = line.slice("rename to ".length)
      lines[i] = "rename from " + toName
      lines[++i] = "rename to " + fromName
    } else if (line.startsWith("@@ ")) {
      const result = parseHunkHeaderLine(line)
      if (result.error) {
        return false
      }
      const { original, patched } = result
      lines[
        i
      ] = `@@ -${patched.start},${patched.length} +${original.start},${original.length} @@`
    } else if (line.startsWith("-")) {
      lines[i] = "+" + line.slice(1)
    } else if (line.startsWith("+")) {
      lines[i] = "-" + line.slice(1)
    }
  }
  return true
}

type Effect =
  | { type: "delete"; path: string }
  | { type: "rename"; fromPath: string; toPath: string }
  | { type: "create"; path: string; contents: string }

export function executeEffects(effects: Effect[]) {
  effects.forEach(eff => {
    switch (eff.type) {
      case "delete":
        fs.unlinkSync(eff.path)
        break
      case "rename":
        fs.moveSync(eff.fromPath, eff.toPath)
        break
      case "create":
        fs.writeFileSync(eff.path, eff.contents)
        break
    }
  })
}

type PatchResult =
  | {
      error: false
      effects: Effect[]
    }
  | {
      error: true
      message?: string
    }

export function patch(
  patchFileContents: string,
  {
    reverse = false,
  }: {
    reverse?: boolean
  } = {},
): PatchResult {
  const effects: Effect[] = []
  // invert patch file here first

  const patchFileLines = patchFileContents.split(/\r?\n/)

  if (reverse && !invertPatchFileLines(patchFileLines)) {
    return {
      error: true,
      message: "failed to reverse patch file",
    }
  }

  const moves: Record<string, string> = {}

  let i = 0
  while (i < patchFileLines.length) {
    const line = patchFileLines[i++].trim()
    if (!line.startsWith("---") && !line.startsWith("rename from")) {
      continue
    }

    if (line.startsWith("rename from")) {
      const fromPath = line.slice("rename from ".length)
      const toPath = patchFileLines[i++].slice("rename to ".length).trim()
      moves[fromPath] = toPath
      effects.push({ type: "rename", fromPath, toPath })
      continue
    }

    const startPath = line.slice("--- ".length)
    const endPath = patchFileLines[i++].trim().slice("--- ".length)

    if (endPath === "/dev/null") {
      // deleting a file
      // just get rid of that noise
      // slice of the 'a/'
      effects.push({
        type: "delete",
        path: startPath.slice(2),
      })
      // ignore hunk header
      const result = parseHunkHeaderLine(patchFileLines[i++])
      if (result.error) {
        return result
      }
      // ignore all -lines
      // TODO: check that the lines match the file
      while (i < patchFileLines.length && patchFileLines[i].startsWith("-")) {
        i++
      }
    } else if (startPath === "/dev/null") {
      // creating a new file
      // just grab all the contents and put it in the file
      const result = parseHunkHeaderLine(patchFileLines[i++])
      if (result.error) {
        return result
      }
      const { patched: { length } } = result
      const fileLines = []
      while (i < patchFileLines.length && patchFileLines[i].startsWith("+")) {
        fileLines.push(patchFileLines[i++].slice(1))
      }
      if (fileLines.length !== length) {
        console.warn(
          "hunk length mismatch :( expected",
          length,
          "got",
          fileLines.length,
        )
      }
      effects.push({
        type: "create",
        path: endPath.slice(2),
        contents: fileLines.join("\n"),
      })
    } else {
      // modifying the file in place
      // check to see if the file has moved first
      const startPathCanonical =
        moves[startPath.slice(2).trim()] || startPath.slice(2)
      const fileLines = fs
        .readFileSync(startPathCanonical)
        .toString()
        .split(/\r?\n/)

      // iterate over hunks
      while (i < patchFileLines.length && patchFileLines[i].startsWith("@@")) {
        const hunkHeader = parseHunkHeaderLine(patchFileLines[i++])
        if (hunkHeader.error) {
          return hunkHeader
        }

        // contextIndex is the offest from the hunk header start but in the original file
        let contextIndex = 0
        while (
          i < patchFileLines.length &&
          patchFileLines[i].match(/^(\+|-| ).*/)
        ) {
          while (
            i < patchFileLines.length &&
            patchFileLines[i].startsWith(" ")
          ) {
            // context line
            // check it's the same as in the file
            const fromPatchFile = trimRight(patchFileLines[i++].slice(1))
            const fromActualFile = trimRight(
              fileLines[hunkHeader.original.start - 1 + contextIndex++],
            )
            if (fromPatchFile !== fromActualFile) {
              return {
                error: true,
                message: `mismatched lines \n  ${JSON.stringify(
                  fromPatchFile,
                )} \n  ${JSON.stringify(fromActualFile)}`,
              }
            }
          }
          while (
            i < patchFileLines.length &&
            patchFileLines[i].startsWith("-")
          ) {
            // delete the line
            // check it's the same as in the file
            if (
              trimRight(patchFileLines[i++].slice(1)) !==
              trimRight(fileLines[hunkHeader.original.start - 1 + contextIndex])
            ) {
              return {
                error: true,
                message: `mismatched lines '${patchFileLines[i - 1].slice(
                  1,
                )}' and '${fileLines[
                  hunkHeader.original.start - 1 + contextIndex
                ]}'`,
              }
            }
            // all good then delete the shizz
            fileLines.splice(hunkHeader.original.start - 1 + contextIndex, 1)
          }
          while (
            i < patchFileLines.length &&
            patchFileLines[i].startsWith("+")
          ) {
            // insert the line
            // check it's the same as in the file
            fileLines.splice(
              hunkHeader.original.start - 1 + contextIndex++,
              0,
              patchFileLines[i++].slice(1),
            )
          }
        }
      }

      // all done with this file?
      effects.push(
        {
          type: "delete",
          path: startPathCanonical,
        },
        {
          type: "create",
          path: endPath.slice(2),
          contents: fileLines.join("\n"),
        },
      )
    }
  }

  return { error: false, effects }
}
