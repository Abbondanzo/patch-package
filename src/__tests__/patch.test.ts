// tslint:disable

import { patch } from "../patch"
import * as fs from "fs-extra"
import * as path from "path"

const properReadFileSync = fs.readFileSync
const properWriteFileSync = fs.writeFileSync
const properUnlinkSync = fs.unlinkSync
const properMoveSync = fs.moveSync

function getMockData() {
  const files = fs.readdirSync(
    path.join(__dirname, "patch-algorithm-test-files"),
  )

  const mockFs: { [fileName: string]: string } = {}
  const patches: { fileName: string; patchContents: string }[] = []

  files.forEach(fileName => {
    if (fileName.endsWith(".patch")) {
      patches.push({
        fileName,
        patchContents: fs
          .readFileSync(
            path.join(__dirname, "patch-algorithm-test-files", fileName),
          )
          .toString(),
      })
    } else {
      mockFs[fileName] = fs
        .readFileSync(
          path.join(__dirname, "patch-algorithm-test-files", fileName),
        )
        .toString()
    }
  })

  return { mockFs, patches }
}

const DATA = Object.freeze(getMockData())

beforeEach(() => {
  // tslint:disable
  ;(fs as any).readFileSync = jest.fn(path => {
    return DATA.mockFs[path]
  })
})

afterEach(() => {
  ;(fs as any).readFileSync = properReadFileSync
  ;(fs as any).writeFileSync = properWriteFileSync
  ;(fs as any).unlinkSync = properUnlinkSync
  ;(fs as any).moveSync = properMoveSync
})

DATA.patches.forEach(({ fileName, patchContents }) => {
  describe(fileName, () => {
    it("can be applied", () => {
      const result = patch(patchContents)
      expect(result.error).toBeFalsy()

      if (result.error) return

      expect(result.effects).toMatchSnapshot()
    })
  })
})
