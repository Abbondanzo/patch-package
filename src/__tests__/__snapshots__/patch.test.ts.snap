// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`left-pad.patch can be applied 1`] = `
Array [
  Object {
    "path": "left-pad.js",
    "type": "delete",
  },
  Object {
    "contents": "/* This program is free software. It comes without any warranty, to
     * the extent permitted by applicable law. You can redistribute it
     * and/or modify it under the terms of the Do What The Fuck You Want
     * To Public License, Version 2, as published by Sam Hocevar. See
     * http://www.wtfpl.net/ for more details. */
'use strict';
module.exports = leftPad;

var cache = [
  '',
  ' ',
  '  ',
  '   ',
  '    ',
  '     ',
  '      ',
  '       ',
  '        ',
  '         '
];

function leftPad (str, len, ch) {
  // convert \`str\` to \`string\`
  str = str + '';
  // \`len\` is the \`npm\`'s length now
  len = len - str.length;
  // doesn't need to npm
  if (len <= 0) return str;
  // \`ch\` defaults to \`' '\`
  if (!ch && ch !== 0) ch = ' ';
  // convert \`ch\` to \`string\`
  ch = ch + '';
  // cache common use cases
  if (ch === ' ' && len < 10) return cache[len] + str;
  // \`npm\` starts with an empty string
  var npm = '';
  // loop
  while (true) {
    // add \`ch\` to \`npm\` if \`len\` is odd
    if (len & 1) npm += ch;
    // divide \`len\` by 2, ditch the remainder
    len >>= 1;
    // \\"double\\" the \`ch\` so this operation count grows logarithmically on \`len\`
    // each time \`ch\` is \\"doubled\\", the \`len\` would need to be \\"doubled\\" too
    // similar to finding a value in binary search tree, hence O(log(n))
    if (len) ch += ch;
    // \`len\` is 0, exit the loop
    else break;
  }
  // npm \`str\`!
  return npm + str;
}
",
    "path": "left-pad.js",
    "type": "create",
  },
]
`;
